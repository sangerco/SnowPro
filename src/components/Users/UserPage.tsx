import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  Dimmer,
  Loader,
  Grid,
  Container,
  Header,
  Image,
  Modal,
  Button,
  Icon,
  Card,
  Divider,
  Embed,
} from "semantic-ui-react";
import { Link, useParams } from "react-router-dom";
import { RootState, AppDispatch } from "../../redux/store";
import { fetchOneUser, deleteUser } from "../../redux/slices/userSlice";
import Inbox from "../Messages/Inbox";
import FavMountain from "../SkiAreas/FavMountain";

const UserPage: React.FC = () => {
  const auth = useSelector((state: RootState) => state.auth);
  const authId = auth.data?.id;
  const { username } = useParams();
  const users = useSelector((state: RootState) => state.users);
  const user = users.user;
  console.log(user?.favMountains);
  const dispatch = useDispatch<AppDispatch>();

  const [showDeleteModal, setShowDeleteModal] = useState(false);

  useEffect(() => {
    if (username) {
      dispatch(fetchOneUser(username));
    }
  }, [dispatch, username]);

  const handleShowDeleteModal = () => {
    setShowDeleteModal(true);
  };

  const handleDeleteUser = () => {
    if (username) {
      dispatch(deleteUser(username));
    }
  };

  let videoIds = [];

  if (user && user.videos && user.videos.length > 0) {
    for (let i = 0; i < user.videos.length; i++) {
      let splitLink = user.videos[i].split("v=");
      let id = splitLink[1];
      let videoId = id.includes("&") ? id.split("&")[0] : id;
      videoIds.push(videoId);
    }
  }

  if (username && user?.favMountains && user.favMountains.length > 0) {
    console.log("true");
  }

  if (users.loading) {
    return (
      <div>
        <Dimmer active>
          <Loader>Loading...</Loader>
        </Dimmer>
      </div>
    );
  }

  if (users.error) {
    return <div>Error: {users.error}</div>;
  }

  if (user) {
    return (
      <div>
        <Container>
          <Grid relaxed>
            <Grid.Row>
              <Grid.Column width={4}>
                <Image src={user.avatar} size="small" />
              </Grid.Column>
              <Grid.Column width={8}>
                <Header as="h2" textAlign="right">
                  {user.username}
                </Header>
                <Header.Subheader>
                  {user.firstName && user.firstName !== ""
                    ? user.firstName
                    : ""}{" "}
                  {user.lastName && user.lastName !== "" ? user.lastName : ""}
                </Header.Subheader>
                {authId && authId === user.id ? (
                  <div>
                    <Link to={`/users/${user.username}/update`}>
                      <Icon name="edit" style={{ cursor: "pointer" }} />
                    </Link>
                    <Icon
                      name="trash"
                      style={{ cursor: "pointer" }}
                      onClick={handleShowDeleteModal}
                    />
                  </div>
                ) : null}
              </Grid.Column>
              <Grid.Column width={4}></Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column width={4}></Grid.Column>
              <Grid.Column width={8}>
                <p>{user.bio}</p>
              </Grid.Column>
              <Grid.Column width={4}>
                <Header as="h3">Favorite Mountains</Header>
                <Divider />
                {user.favMountains &&
                username &&
                user.favMountains.length > 0 ? (
                  user.favMountains.map((fm) => (
                    <Card.Group>
                      <FavMountain key={fm} slug={fm} username={username} />
                    </Card.Group>
                  ))
                ) : (
                  <p>This user has no favorited mountains yet!</p>
                )}
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column width={2}></Grid.Column>
              <Grid.Column width={12}>
                <Header as="h3">
                  <Link to={`/users/${username}/photos`}>
                    All {username}'s Photos
                  </Link>
                </Header>
                {user.photos && user.photos.length > 0 ? (
                  user.photos.map((photo) => <Image size="small" src={photo} />)
                ) : (
                  <p>This user has no photos yet!</p>
                )}
                <Divider />
              </Grid.Column>
              <Grid.Column width={2}></Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column width={2}></Grid.Column>
              <Grid.Column width={12}>
                <Header as="h3">
                  <Link to={`/users/${username}/videos`}>
                    All {username}'s Videos
                  </Link>
                </Header>
                {videoIds && videoIds.length > 0 ? (
                  videoIds.map((video) => (
                    <Embed size="small" id={video} source="youtube" />
                  ))
                ) : (
                  <p>This user has no videos yet!</p>
                )}
                <Divider />
              </Grid.Column>
              <Grid.Column width={2}></Grid.Column>
            </Grid.Row>
          </Grid>
        </Container>
        <Modal open={showDeleteModal} onClose={() => setShowDeleteModal(false)}>
          <Modal.Content>
            Are You Sure You Want To Delete Your Profile?
          </Modal.Content>
          <Modal.Actions>
            <Button negative onClick={handleDeleteUser}>
              Yes
            </Button>
            <Button onClick={() => setShowDeleteModal(false)}>Cancel</Button>
          </Modal.Actions>
        </Modal>
      </div>
    );
  }

  return null;
};

export default UserPage;
