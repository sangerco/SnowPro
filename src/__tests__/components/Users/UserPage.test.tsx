/* eslint-disable testing-library/no-node-access */
import React from "react";
import UserPage from "../../../components/Users/UserPage";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import nock from "nock";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test User Page", () => {
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    reviews: {
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review",
          body: "test review body",
          stars: 2,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
      ],
    },
    reviewReplies: {
      reviewReply: {
        id: "review-reply-1",
        reviewId: "review-1",
        userId: "22",
        username: "testuser2",
        body: "test review reply body",
        slug: "ski-area-1",
        createdAt: "fake created date",
      },
    },
    skiAreas: {
      skiArea: {
        slug: "ski-area-1",
        name: "Ski Area 1",
        country: "US",
        region: "CO",
        href: "www.fakehref.com",
        location: {
          latitude: 0,
          longitude: 0,
        },
        lifts: {},
        twitter: {},
      },
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
    users: {
      user: {
        id: "11",
        username: "testuser",
        firstName: "Test",
        lastName: "User",
        email: "test@user.com",
        bio: "test user bio",
        avatar: "https://testdata.com/testavatar.jpg",
        isAdmin: false,
      },
      users: [
        {
          id: "11",
          username: "testuser",
          firstName: "Test",
          lastName: "User",
          email: "test@user.com",
          bio: "test user bio",
          avatar: "https://testdata.com/testavatar.jpg",
          isAdmin: false,
        },
        {
          id: "22",
          username: "testuser2",
          firstName: "Test",
          lastName: "User2",
          email: "test@user2.com",
          bio: "test user bio",
          avatar: "https://testdata.com/testavatar2.jpg",
          isAdmin: true,
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UserPage />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <UserPage />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

describe("test show user", () => {
  const username = "testuser";
  const useParamsMock = jest.requireMock("react-router-dom")
    .useParams as jest.Mock;
  useParamsMock.mockReturnValue({ username });

  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    reviews: {
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review",
          body: "test review body",
          stars: 2,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
      ],
    },
    reviewReplies: {
      reviewReply: {
        id: "review-reply-1",
        reviewId: "review-1",
        userId: "22",
        username: "testuser2",
        body: "test review reply body",
        slug: "ski-area-1",
        createdAt: "fake created date",
      },
    },
    skiAreas: {
      skiArea: {
        slug: "ski-area-1",
        name: "Ski Area 1",
        country: "US",
        region: "CO",
        href: "www.fakehref.com",
        location: {
          latitude: 0,
          longitude: 0,
        },
        lifts: {},
        twitter: {},
      },
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
    users: {
      user: {
        id: "11",
        username: "testuser",
        firstName: "Test",
        lastName: "User",
        email: "test@user.com",
        bio: "test user bio",
        avatar: "https://testdata.com/testavatar.jpg",
        isAdmin: false,
        videos: [],
        photos: [],
        favMountains: [],
      },
      users: [
        {
          id: "11",
          username: "testuser",
          firstName: "Test",
          lastName: "User",
          email: "test@user.com",
          bio: "test user bio",
          avatar: "https://testdata.com/testavatar.jpg",
          isAdmin: false,
        },
        {
          id: "22",
          username: "testuser2",
          firstName: "Test",
          lastName: "User2",
          email: "test@user2.com",
          bio: "test user bio",
          avatar: "https://testdata.com/testavatar2.jpg",
          isAdmin: true,
        },
      ],
    },
  };
  const store = mockStore(initialState);

  it("should display a user", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get(`/users/${username}`)
      .reply(200, {
        id: "11",
        username: "testuser",
        firstName: "Test",
        lastName: "User",
        email: "test@user.com",
        password: "password",
        avatar: "https://testlinks.com/testavatar.jpg",
        bio: "test user bio",
        isAdmin: "false",
        videos: [],
        photos: [],
        favMountains: [],
      });
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UserPage />
        </BrowserRouter>
      </Provider>
    );

    await waitFor(() => {
      const username = screen.getByTestId("user-header");
      expect(username).toBeInTheDocument();
    });

    await waitFor(() => {
      const name = screen.getByTestId("user-name");
      expect(name).toBeInTheDocument();
    });
  });

  it("should delete button and delete modal", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get(`/users/${username}`)
      .reply(200, {
        id: "11",
        username: "testuser",
        firstName: "Test",
        lastName: "User",
        email: "test@user.com",
        password: "password",
        avatar: "https://testlinks.com/testavatar.jpg",
        bio: "test user bio",
        isAdmin: "false",
        videos: [],
        photos: [],
        favMountains: [],
      });
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UserPage />
        </BrowserRouter>
      </Provider>
    );

    const deleteButton = screen.getByTestId("delete-button");
    fireEvent.click(deleteButton);

    await waitFor(() => {
      const modal = screen.getByTestId("delete-modal");
      expect(modal).toBeInTheDocument();
    });
  });
});
