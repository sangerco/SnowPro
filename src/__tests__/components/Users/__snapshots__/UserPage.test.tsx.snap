// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test User Page should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="ui container"
    >
      <div
        class="ui relaxed grid"
      >
        <div
          class="row"
        >
          <div
            class="four wide column"
          >
            <img
              class="ui small image"
              src="https://testdata.com/testavatar.jpg"
            />
          </div>
          <div
            class="eight wide column"
          >
            <h2
              class="ui right aligned header"
              data-testid="user-header"
            >
              testuser
            </h2>
            <div
              class="sub header"
              data-testid="user-name"
            >
              Test User
            </div>
            <div>
              <a
                href="/users/testuser/update"
              >
                <i
                  aria-hidden="true"
                  class="edit icon"
                  style="cursor: pointer;"
                />
              </a>
              <i
                aria-hidden="true"
                class="trash icon"
                data-testid="delete-button"
                style="cursor: pointer;"
              />
            </div>
          </div>
          <div
            class="four wide column"
          />
        </div>
        <div
          class="row"
        >
          <div
            class="four wide column"
          />
          <div
            class="eight wide column"
          >
            <p>
              test user bio
            </p>
          </div>
          <div
            class="four wide column"
          >
            <h3
              class="ui header"
            >
              Favorite Mountains
            </h3>
            <div
              class="ui divider"
            />
            <p>
              This user has no favorited mountains yet!
            </p>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="two wide column"
          />
          <div
            class="twelve wide column"
          >
            <h3
              class="ui header"
            >
              <a
                href="/users/testuser/photos"
              >
                All testuser's Photos
              </a>
            </h3>
            <p>
              This user has no photos yet!
            </p>
            <div
              class="ui divider"
            />
          </div>
          <div
            class="two wide column"
          />
        </div>
        <div
          class="row"
        >
          <div
            class="two wide column"
          />
          <div
            class="twelve wide column"
          >
            <h3
              class="ui header"
            >
              <a
                href="/users/testuser/videos"
              >
                All testuser's Videos
              </a>
            </h3>
            <p>
              This user has no videos yet!
            </p>
            <div
              class="ui divider"
            />
          </div>
          <div
            class="two wide column"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
